packages/tools/devtools/devtools-test-app/src/widgets/TodoItemView.tsx

```diff
@@ -0,0 +1,162 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { makeStyles } from "@fluentui/react-components";
+import {
+	CollaborativeInput,
+	CollaborativeTextArea,
+	SharedStringHelper,
+} from "@fluid-example/example-utils";
+import type { SharedString } from "@fluidframework/sequence/internal";
+import { Tree, type TreeNode } from "@fluidframework/tree/internal";
+import React from "react";
+
+import type { TodoItem } from "../Schema.js";
+
+const useStyles = makeStyles({
+	todoItemHeader: {
+		margin: 0,
+		display: "flex",
+	},
+	todoItemExpandButton: {
+		boxSizing: "border-box",
+		width: "30px",
+		height: "50px",
+		margin: "10px 0 10px 10px",
+		border: "none",
+		padding: 0,
+		fontSize: "30px",
+		background: "none",
+	},
+	todoItemCheckbox: {
+		boxSizing: "border-box",
+		width: "50px",
+		height: "50px",
+		margin: "10px 0 10px 10px",
+	},
+	todoItemInput: {
+		boxSizing: "border-box",
+		border: "1px solid #666",
+		width: "100%",
+		height: "50px",
+		margin: "10px 0 10px 10px",
+		padding: 0,
+		fontSize: "20px",
+		outline: "none",
+	},
+	todoItemDetails: {
+		width: "100%",
+	},
+});
+
+/**
+ * {@link TodoItemView} input props.
+ * @internal
+ */
+interface TodoItemViewProps {
+	readonly todoItemModel: TodoItem;
+	readonly className?: string;
+}
+
+/**
+ * Todo Item that will be stored in the {@link TodoListView} tree.
+ * Contains a title, description and a checkbox to mark it as completed.
+ */
+export const TodoItemView: React.FC<TodoItemViewProps> = (props: TodoItemViewProps) => {
+	const { todoItemModel, className } = props;
+	const styles = useStyles();
+
+	const [itemTitle, setItemTitle] = React.useState<SharedString | undefined>(undefined);
+	const [itemDescription, setItemDescription] = React.useState<SharedString | undefined>(
+		undefined,
+	);
+	const [detailsVisible, setDetailsVisible] = React.useState<boolean>(false);
+
+	useTree(todoItemModel);
+
+	React.useEffect(() => {
+		todoItemModel.title
+			.get()
+			.then((text) => {
+				setItemTitle(text as SharedString);
+			})
+			.catch((error) => {
+				console.error("Failed to load title:", error);
+			});
+	}, [todoItemModel.title]);
+
+	React.useEffect(() => {
+		todoItemModel.description
+			.get()
+			.then((text) => {
+				setItemDescription(text as SharedString);
+			})
+			.catch((error) => {
+				console.error("Failed to load description:", error);
+			});
+	}, [todoItemModel.description]);
```

Comment:
This feels funky, and it causes the component to render twice each time it loads (once after each of the two fields is "resolved"). Maybe try something like this inside a single effect?

```typescript
const titleP = todoItemModel.title.get();
const descriptionP = todoItemModel.description.get();
Promise.all([titleP, descriptionP]).then([title, description] => {
  setItemTitle(title as SharedString);
  setItemDescription(description as SharedString);
})
.catch((error) => {
  console.error("Failed to load title or description:", error);
});
```

Reply:
I believe keeping the two seprate `useEffect` hooks would rather avoid the double render problem? Integrating everything into a single hook also looks cleaner, but the [original todo‑app](https://github.com/microsoft/FluidFramework/blob/52ba8aa3d0539c4eaf18fbf9c7e30c65cae64536/examples/data-objects/todo/src/TodoItem/TodoItemView.tsx#L24) seems to have used two separate hooks so the component doesn’t rerender when only title or description changes.

If we collapse everything into one hook that resolves a single promise, we’d still have to fetch both title and description even when just one of them updates.

Reply:
I agree with Ji - I think the separate `useEffect`s is better and more idiomatic. Consider the case where the `description` field is updated with a new `SharedString` handle. In that case, we want to re-run the necessary update logic for the new `description`, but there is no need to re-run the logic for `title`.

Reply:
👍 no objections